### PR TITLE
[SofaCarving] Refresh and enhancement

### PR DESCRIPTION
**Pull request by epernod**
- Fix and add more examples
- Add a carvingDistance parameter to be able to have real collision before carving.
- Add the possiblity to handle several collision models on the target object.
- Clean and Doc the code.

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-scene-tests]

**Comment by epernod** [comment]
Results movies:
<a href="https://www.youtube.com/watch?v=-lnQ1oo4JJ8" target="_blank">movie1</a>
<a href="https://youtu.be/vLRVZR9kwuY" target="_blank">movie2</a>

**Comment by epernod** [comment]
https://www.youtube.com/watch?v=vLRVZR9kwuY

**Comment by epernod** [comment]
[ci-build][with-scene-tests]

**Comment by guparan** [comment]
Guidelines: https://github.com/sofa-framework/sofa/blob/master/GUIDELINES.md

File: applications/plugins/SofaCarving/CarvingManager.cpp [hunk 1]
```diff
@@ -50,178 +49,209 @@ int CarvingManagerClass = core::RegisterObject("Manager handling carving operati
...
-    
     sofa::helper::vector<std::pair<core::CollisionModel*, core::CollisionModel*> > vectCMPair;
-    vectCMPair.push_back(std::make_pair(modelSurface->getFirst(),modelTool->getFirst()));
-    
-    detectionNP->setInstance(this);
-    detectionNP->setIntersectionMethod(intersectionMethod);
-    detectionNP->beginNarrowPhase();
-    detectionNP->addCollisionPairs(vectCMPair);
-    detectionNP->endNarrowPhase();
+    for (int i = 0; i < m_surfaceCollisionModels.size(); i++)
```
Comment (marques-bruno) [inline applications/plugins/SofaCarving/CarvingManager.cpp]: size_t ?

Comment (epernod) [reply applications/plugins/SofaCarving/CarvingManager.cpp]: ......... 👍

**Pull request merged**